### PR TITLE
Allow jobtype implementors to set task progress

### PR DESCRIPTION
**Pull request by guidow**

### CODE REVIEW

File: pyfarm/jobtypes/core/jobtype.py [hunk 1]
```diff
@@ -789,6 +794,99 @@ def set_states(self, tasks, state, error=None):
...
+                else:
+                    logger.debug("While posting progress update for task %s to "
+                                 "master, caught %s. Retrying immediately.",
+                                 task["id"], error.__class__.__name__)
+            except Exception as error:
+                message = (
+                    "Failed to post progress update for task %s to the master: "
+                    "%r." % (task["id"], error))
+                logger.error(message)
+                raise Exception(message)
```
Comment (opalmer) [inline pyfarm/jobtypes/core/jobtype.py]: You probably should just do `raise` here so we keep the traceback.

File: pyfarm/jobtypes/core/jobtype.py [hunk 2]
```diff
@@ -789,6 +794,99 @@ def set_states(self, tasks, state, error=None):
...
+                        "not be retried." % (task["id"], response_data))
+                    logger.error(message)
+                    raise Exception(message)
+
+                else:
+                    message = (
+                        "Unhandled error when posting progress update for task "
+                        "%s to the master: %s (code: %s).  This request will "
+                        "not be retried." % (response_data, response.code))
+                    logger.error(message)
```
Comment (opalmer) [inline pyfarm/jobtypes/core/jobtype.py]: `logger.error` accepts positional arguments:

```python
logger.error("foo %s", "bar")
```

File: pyfarm/jobtypes/core/jobtype.py [hunk 3]
```diff
@@ -789,6 +794,99 @@ def set_states(self, tasks, state, error=None):
...
+                    raise Exception(message)
+
+            except (ResponseNeverReceived, RequestTransmissionFailed) as error:
+                num_retry_errors += 1
+                if num_retry_errors > config["broken_connection_max_retry"]:
+                    message = (
+                        "Failed to post progress update for task %s to the "
+                        "master, caught try-again type errors %s times in a "
+                        "row." % (task["id"], num_retry_errors))
+                    logger.error(message)
```
Comment (opalmer) [inline pyfarm/jobtypes/core/jobtype.py]: `logger.error` accepts positional arguments:

```python
logger.error("foo %s", "bar")
```

File: pyfarm/jobtypes/core/jobtype.py [hunk 4]
```diff
@@ -789,6 +794,99 @@ def set_states(self, tasks, state, error=None):
...
+                        "in %s seconds.", task["id"], response_data, delay)
+
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                elif response.code == NOT_FOUND:
+                    message = ("Got 404 NOT FOUND error on updating progress "
+                               "for task %s" % task["id"])
+                    logger.error(message)
```
Comment (opalmer) [inline pyfarm/jobtypes/core/jobtype.py]: `logger.error` accepts positional arguments:

```python
logger.error("foo %s", "bar")
```

File: pyfarm/jobtypes/core/jobtype.py (hunk 2 above)
Comment (guidow) [inline pyfarm/jobtypes/core/jobtype.py]: Yes, but I am using message twice, so I'm constructing it once beforehand.

File: pyfarm/jobtypes/core/jobtype.py (hunk 3 above)
Comment (guidow) [inline pyfarm/jobtypes/core/jobtype.py]: Yes, but I am using message twice, so I'm constructing it once beforehand.

File: pyfarm/jobtypes/core/jobtype.py (hunk 4 above)
Comment (guidow) [inline pyfarm/jobtypes/core/jobtype.py]: Yes, but I am using message twice, so I'm constructing it once beforehand.

**Comment by opalmer** [comment]
Going to go ahead and merge this, I don't see any reasons to hold this back.

**Pull request merged**